test/extensions/filters/http/oauth2/oauth_integration_test.cc

```diff
@@ -211,8 +297,108 @@ name: oauth
+  response->waitForHeaders();
+  EXPECT_EQ("302", response->headers().getStatusValue());
+}
+
+TEST_P(OauthIntegrationTest, AuthenticationFlow) {
+  on_server_init_function_ = [&]() {
+    createLdsStream();
+    sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "initial");
+  };
+
+  initialize();
+
+  // 1. Do one authentication flow.
+  doAuthenticationFlow("token_secret", "hmac_secret");
+
+  // 2. Reload secrets.
+  EXPECT_EQ(test_server_->counter("sds.token.update_success")->value(), 1);
+  EXPECT_EQ(test_server_->counter("sds.hmac.update_success")->value(), 1);
+  TestEnvironment::renameFile(TestEnvironment::temporaryPath("token_secret_1.yaml"),
+                              TestEnvironment::temporaryPath("token_secret.yaml"));
+  test_server_->waitForCounterEq("sds.token.update_success", 2, std::chrono::milliseconds(5000));
+  TestEnvironment::renameFile(TestEnvironment::temporaryPath("hmac_secret_1.yaml"),
+                              TestEnvironment::temporaryPath("hmac_secret.yaml"));
+  test_server_->waitForCounterEq("sds.hmac.update_success", 2, std::chrono::milliseconds(5000));
+  // 3. Do another one authentication flow.
+  doAuthenticationFlow("token_secret_1", "hmac_secret_1");
+}
+
+// Regression test(issue #22678) where (incorrectly)using server's init manager(initialized state)
+// to add init target by the secret manager led to the assertion failure.
+TEST_P(OauthIntegrationTest, LoadListenerAfterServerIsInitialized) {
+  on_server_init_function_ = [&]() {
+    createLdsStream();
+    envoy::config::listener::v3::Listener listener =
+        TestUtility::parseYaml<envoy::config::listener::v3::Listener>(R"EOF(
+      name: fake_listener
+      address:
+        socket_address:
+          address: 127.0.0.1
+          port_value: 0
+      filter_chains:
+        - filters:
+          - name: envoy.filters.network.http_connection_manager
+            typed_config:
+              "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
+              codec_type: HTTP2
+              stat_prefix: config_test
+              route_config:
+                name: route_config_0
+                virtual_hosts:
+                  - name: integration
+                    domains:
+                      - "*"
+                    routes:
+                      - match:
+                          prefix: /
+                        route:
+                          cluster: cluster_0
+              http_filters:
+                - name: envoy.filters.http.router
+        )EOF");
+
+    // dummy listener is being sent so that lds api gets marked as ready, which would
+    // led to server's init manager reach initialized state
+    sendLdsResponse({listener}, "initial");
+  };
+
+  initialize();
+
+  // add listener with oauth2 filter and sds configs
+  sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "delayed");
+
```

Comment:
Since you're going to have to push another commit: I think we need to add to wait for the new listener config to be applied here: `test_server_->waitForCounterGe("listener_manager.lds.update_success", 2);`.

Reply:
Also: `test_server_->waitForGaugeEq("listener_manager.total_listeners_warming", 0);`

Reply:
done!